packages/api/src/controllers/transcode.ts

```diff
@@ -0,0 +1,54 @@
+import { authorizer, validatePost } from "../middleware";
+import { Router } from "express";
+import { TranscodePayload } from "../schema/types";
+import { toObjectStoreUrl } from "./helpers";
+import { ensureQueueCapacity } from "../task/scheduler";
+
+const app = Router();
+
+app.post(
+  "/",
+  authorizer({}),
+  validatePost("transcode-payload"),
+  async (req, res) => {
+    const params = req.body as TranscodePayload;
+    const { catalystPipelineStrategy = undefined } = req.user.admin
+      ? params
+      : {};
+
+    await ensureQueueCapacity(req.config, req.user.id);
+
+    const taskType = "transcode-file";
+    let inUrl = params.input.url;
+    if (!inUrl) {
+      if (!params.input.path) {
+        throw new Error("Undefined property 'input.path'");
+      }
+      inUrl = toObjectStoreUrl(params.input) + params.input.path;
+    }
```

Comment:
This feels like a JSON schema + "type-switch" kind of thing, WDYT? We could have a logic here like
```suggestion
    let inUrl: string
    switch (params.input.type ?? "url") {
      case "url":
        inUrl = params.input.url
        break
      case "s3":
        inUrl = toObjectStoreUrl(params.input) + params.input.path
        break
      default:
        throw new Error(`Unknown input type: `${params.input.type}`)
   }
   if (!inUrl) {
     throw new Error(`Missing input URL`) // this should never happen, just a safeguard
   }
```

And the JSON schema with a proper `oneOf` definition for the `input field etc, would guarantee more specific errors on diversions from the schema (e.g. type s3 but no creds, path or bucket, etc.)

Reply:
JSON schema is fine. But to keep it simple, I'd probably not add type `url`. I think you can just specify URL and then it's used. That's also how other services do it.